proxy: add a header to be able to disable compression on demand

deweerdt opened the pull request:
This commit introduces a `x-do-not-compress` header that allows an agent
downstream to ask H2O to disable it's compression logic.
This allows to implement on demand compression outside of H2O.

kazuho commented:
Thank you for the PR.

> This allows to implement on demand compression outside of H2O.

Am I correct in assuming that your intention is to _decide_ whether if a content should be compressed outside of H2O, but let H2O do the compression?

The reason I ask is because if you are to decide _and apply_ compression above H2O, there would be no need to turn on H2O's on-the-fly compression module.

And if my assumption is correct, it would be great if the header was defined to take a boolean argument, specifying either to compress or not compress the content. Upon observing the header, the on-the-fly compressor would act as specified. Or if the header is absent, the compressor would use the default rules (i.e. mime-type based) to decide whether or not to compress the response.

deweerdt commented:
Yes that is correct, the goal is to be able to influence the decision that H2O makes. The reason I went with an opt out header is that making it a boolean felt a bit awkward in the 'on' case: `x-compress:on` would only work if compression was activated in the config and the minimum size matched and the `accept-encoding` header fit our available encodings.
OTOH, do not compress always does what it says, no ambiguity.

deweerdt commented:
I'm realizing the change you're proposing is really about what this header control. I viewed the change as something to be done on top of H2O's default configuration mechanism, not something that would override it.
If I understand you correctly, the header would have no effect if `compress:OFF`  was set, but i all the other cases (including the minimum size), the header would take precedence. Is that what you are suggesting?

deweerdt commented:
I've implemented something that i believe is closer to what you suggested, and added a test in db65f0f

kazuho commented:
> I'm realizing the change you're proposing is really about what this header control. I viewed the change as something to be done on top of H2O's default configuration mechanism, not something that would override it.
> If I understand you correctly, the header would have no effect if `compress:OFF` was set, but i all the other cases (including the minimum size), the header would take precedence. Is that what you are suggesting?

Yes. That was my intention, though _how_ it would override the configuration could be subject to further discussion.

My idea has been that an upstream server should be able to specify _how_ it wants the content to be compressed (and when specified, value of the header should override the configuration for the compress handler found in the YAML file). For example, the header might look like:
```
x-compress: off             # do not compress
x-compress: auto            # apply the rules found in the configuration
x-compress: on              # compress the data
x-compress: gzip; quality=4 # use gzip compression (with the quality parameter set to 4)
```

kazuho commented on lib/core/proxy.c:
```diff
@@ -421,6 +421,10 @@ static h2o_http1client_body_cb on_head(h2o_http1client_t *client, const char *er
                 h2o_push_path_in_link_header(req, headers[i].value, headers[i].value_len);
+            } else if (token == H2O_TOKEN_X_COMPRESS) {
+                req->compression_hint =
+                    h2o_strtosize(headers[i].value, headers[i].value_len) ? H2O_COMPRESS_HINT_ENABLE : H2O_COMPRESS_HINT_DISABLE;
```
Could we accept caseless strings (e.g. `off`, `on`, `auto`) instead of numbers to indicate the mode?
Of course, we should use an enum internally (as we already do).

kazuho commented on lib/handler/compress.c:
```diff
@@ -61,12 +61,19 @@ static void on_setup_ostream(h2o_filter_t *_self, h2o_req_t *req, h2o_ostream_t
-        goto Next;
-    if (req->res.content_length < self->args.min_size)
+    /* compression was explicitely disabled, skip */
+    if (req->compression_hint == H2O_COMPRESS_HINT_DISABLE)
```
For readability, how about using a switch statement covering all possible values taken by the compression hint (e.g. `switch (req->compression_hint)`)?

kazuho commented on lib/handler/compress.c:
```diff
@@ -61,12 +61,19 @@ static void on_setup_ostream(h2o_filter_t *_self, h2o_req_t *req, h2o_ostream_t
+        if (!req->res.mime_attr->is_compressible)
+            goto Next;
+        if (req->res.content_length < self->args.min_size)
+            goto Next;
```
IMO it might make sense to still _not_ compress the content when the length is below the threshold, even if `x-compress: on` was set?

Otherwise, every upstream configuration that emits the `x-compress` would need to implement the size check. Is that preferable for you?

deweerdt replied:
It is preferable for us that the backend makes the decision, yes.

kazuho replied:
Thank you for the clarification. If that is the case, I agree that we should _not_ check the size of the content when `x-compress: on` is set.

deweerdt commented:
@kazuho thanks for your comments

> x-compress: auto            # apply the rules found in the configuration

If `auto` is equivalent to no header being present, is it still useful?

> x-compress: gzip; quality=4 # use gzip compression (with the quality parameter set to 4)

If the client sent `accept-encoding:brotli` does the above header means we won't compress (because there's no match between the accepted and offered encodings) ?

kazuho commented:
@deweerdt 
>> ```
>> x-compress: auto # apply the rules found in the configuration
>> ```
> If `auto` is equivalent to no header being present, is it still useful?

Though I do not have a strong opinion, having `auto` might help upstream implementers, since their code can always emit the header (having different values depending on the situation), instead of conditionally emitting or not emitting the header.

> If the client sent `accept-encoding:brotli` does the above header means we won't compress (because there's no match between the accepted and offered encodings) ?

That's a good point. It might be a good idea to just implement `off`, `on` (and maybe `auto`) for the time being, and wait until the need for more complex semantics arises.

deweerdt commented:
@kazuho I've implemented the changes you suggested in 179d6c3. In addition to that, i've renamed the header to `compress-hint`, since that might be better for interoperability. Please let me know what you think.

kazuho commented on lib/core/proxy.c:
```diff
@@ -371,6 +371,17 @@ static int on_body(h2o_http1client_t *client, const char *errstr)
 
+static char compress_hint_to_enum(const char *val, size_t len)
+{
+    if (!strncasecmp("ON", val, len)) {
```
Please use `h2o_lcstris` instead of `strncasecmp` so that the comparison would be a complete match (instead of a prefix match; current code would yield true if [val,len] is ["O", 1]) and that it would be locale-independent.

deweerdt replied:
Fixed, thanks.

kazuho commented on misc/tokens.pl:
```diff
@@ -220,3 +220,4 @@ sub normalize_name {
 0 0 0 0 0 x-forwarded-for
 0 0 0 0 0 x-traffic
 0 0 0 0 0 cache-digest
+0 0 0 0 0 compress-hint
```
Should we better have a `x-` prefix?

deweerdt replied:
I removed the `x-` prefix because of https://tools.ietf.org/html/rfc7231.html#section-8.3.1 do you think we should still keep it?

kazuho replied:
Thank you for the link. I think we should retain `x-`, considering the fact that the header is a signal used _internally_ (we drop the header in H2O). The headers that fall into the criteria are not subject to the discussion of deprecation, quoting from RFC 7231:
```
   Authors of specifications defining new fields are advised to keep the
   name as short as practical and not to prefix the name with "X-"
   unless the header field will never be used on the Internet.
```

deweerdt replied:
Makes sense, I've renamed the header. Thanks.

deweerdt commented:
Thanks for your review. I've updated the code to use `h2o_lcstris` instead of `strncasecmp`. Waiting for your response regarding the `x-` prefix.

kazuho commented on lib/core/proxy.c:
```diff
@@ -373,10 +373,10 @@ static int on_body(h2o_http1client_t *client, const char *errstr)
 static char compress_hint_to_enum(const char *val, size_t len)
 {
-    if (!strncasecmp("ON", val, len)) {
+    if (h2o_lcstris(H2O_STRLIT("on"), val, len)) {
```
The order of the argument should be `h2o_lcstris(val, len, H2O_STRLIT("on"))`. The function only converts the first set of arguments to lowercase.

deweerdt replied:
doh, fixed. Thanks

kazuho commented:
Thank you for the updates. Merged to master.
